### PR TITLE
[Improve] read flink on k8s deploy log improvement

### PR DESCRIPTION
**Pull request by wolfboys**
[Improve] read flink on k8s deploy log improvement 

## What changes were proposed in this pull request

1) Removed MoreFeature, added a new class for only one method called , added complexity in my opinion. Needs to be simplified
2) read flink on k8s deploy log optimization
3) Removed LogService, LogServiceImpl 

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

## Does this pull request potentially affect one of the following parts
 - Dependencies (does it add or upgrade a dependency): no

### CODE REVIEW

**Comment by wolfboys** [comment]
cc @MonsterChenzhuo @1996fanrui  PTAL in your free time, thanks~

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java
Comment (1996fanrui) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java]:
```suggestion
      CompletableFuture<String> future =
          CompletableFuture.supplyAsync(
                  () ->
                      KubernetesDeploymentHelper.watchDeploymentLog(
                          application.getK8sNamespace(),
                          application.getJobName(),
                          application.getJobId()))
              .exceptionally(
                  e -> {
                    try {
                      return String.format(
                          "%s/%s_err.log", WebUtils.getAppTempDir(), application.getJobId());
                    } catch (Exception ex) {
                      throw new ApiDetailException(
                          "Generate log path exception: " + ex.getMessage());
                    }
                  })
              .thenApply(path -> logClient.rollViewLog(String.valueOf(path), offset, limit));
```

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java
Comment (1996fanrui) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java]: Should we cancel the future after timeout?

**Review by 1996fanrui — COMMENTED** [review]
Sorry, I'm not familiar with flink on k8s related code, so I just feedback some technical comments.

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java
Comment (wolfboys) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java]: > Should we cancel the future after timeout?

Thanks for your review, this code has been improved and some bugs have been solved, usually the feature does not need to be closed manually.

File: streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java
Comment (1996fanrui) [inline streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java]: All tests should check some conditions instead of `System.out.println`

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/LogClientService.java
Comment (1996fanrui) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/LogClientService.java]: The result can be removed, and return `""` in the end of this method directly.

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java
Comment (1996fanrui) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java]:
```suggestion

      CompletableFuture future = CompletableFuture.supplyAsync(
                  () ->
                      KubernetesDeploymentHelper.watchDeploymentLog(
                          application.getK8sNamespace(),
                          application.getJobName(),
                          application.getJobId()));
      CompletableFutureUtils.supplyTimeout(future,
              5,
              TimeUnit.SECONDS,
              success -> success,
              exception -> {
                String errorLog =
                    String.format(
                        "%s/%s_err.log", WebUtils.getAppTempDir(), application.getJobId());
                File file = new File(errorLog);
                if (file.exists()) {
                  return errorLog;
                } else {
                  throw new ApiDetailException("get k8s job log failed: " + exception.getMessage());
                }
                if (!future.isDone()) {
                  future.cancel(true)
                }
              })
```

The future need to be cancel after timeout.

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/LogClientService.java
Comment (wolfboys) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/LogClientService.java]: > The result can be removed, and return `""` in the end of this method directly.

thanks for your review, done.

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java
Comment (wolfboys) [reply streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/core/service/impl/ApplicationServiceImpl.java]: done

File: streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java
Comment (wolfboys) [reply streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java]: resolved

**Comment by wolfboys** [comment]
cc @MonsterChenzhuo PTAL if your have free time. thanks.

File: streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java
Comment (1996fanrui) [inline streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java:196]:
```suggestion
      resp = exceptionally();
      fail("It should be called.");
```

File: streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java
Comment (1996fanrui) [inline streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java:198]: We should check the exception type here instead of check the `exResult`.

File: streampark-common/src/main/scala/org/apache/streampark/common/util/CompletableFutureUtils.scala
Comment (1996fanrui) [inline streampark-common/src/main/scala/org/apache/streampark/common/util/CompletableFutureUtils.scala]: Check `!future.isDone` first.

File: streampark-common/src/main/scala/org/apache/streampark/common/util/CompletableFutureUtils.scala
Comment (1996fanrui) [inline streampark-common/src/main/scala/org/apache/streampark/common/util/CompletableFutureUtils.scala]: The unitFuture is canceled three times:

image

File: streampark-common/src/main/scala/org/apache/streampark/common/util/CompletableFutureUtils.scala
Comment (1996fanrui) [inline streampark-common/src/main/scala/org/apache/streampark/common/util/CompletableFutureUtils.scala]: Could you please add some unit tests to check the future is canceled after timeout?

**Review by 1996fanrui — CHANGES_REQUESTED** [review]

File: streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java
Comment (1996fanrui) [inline streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java]:
```suggestion

  @Test
  public void thenSupplyException() {
    String expectedExceptionMessage = "Exception in exceptionally handler";
    CompletableFuture<String> future =
        CompletableFutureUtils.supplyTimeout(
            CompletableFuture.supplyAsync(
                () -> {
                  try {
                    Thread.sleep(5000);
                  } catch (InterruptedException e) {
                    throw new RuntimeException(e);
                  }
                  return "success";
                }),
            2,
            TimeUnit.SECONDS,
            success -> success,
            e -> {
              throw new RuntimeException(expectedExceptionMessage);
            });
    assertThatThrownBy(future::get)
        .cause()
        .as(expectedExceptionMessage)
        .isInstanceOf(RuntimeException.class);
  }
```

Using the `assertThatThrownBy` instead of try catch.

**Comment by 1996fanrui** [comment]
Hi @wolfboys , just to managed the expectation, as I said before: `Sorry, I'm not familiar with flink on k8s related code, so I just feedback some technical comments.`

My comments are just code specification, I didn't review the core logical of this PR. So I cannot approve this PR after my comments are resolved. It still needs some k8s reviewers.

cc @MonsterChenzhuo

File: streampark-common/src/main/scala/org/apache/streampark/common/util/CompletableFutureUtils.scala
Comment (wolfboys) [reply streampark-common/src/main/scala/org/apache/streampark/common/util/CompletableFutureUtils.scala]: > Check `!future.isDone` first.

In fact, this judgment is not needed, future.cancel can be executed multiple times, and the judgment is made in the internal implementation.

**Comment by MonsterChenzhuo** [comment]
@wolfboys @1996fanrui  When I run gh login, I can't check pr without a github login jump, so I can't test accordingly. I don't know if it's a problem with the local network or with github

**Comment by wolfboys** [comment]
> @wolfboys @1996fanrui When I run gh login, I can't check pr without a github login jump, so I can't test accordingly. I don't know if it's a problem with the local network or with github

Let me help you, come on

File: streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java
Comment (wolfboys) [reply streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java]: We can catch this exception instead of actually throw this exception, otherwise the test ci verification will fail

File: streampark-common/src/main/scala/org/apache/streampark/common/util/CompletableFutureUtils.scala
Comment (wolfboys) [reply streampark-common/src/main/scala/org/apache/streampark/common/util/CompletableFutureUtils.scala]: > Could you please add some unit tests to check the future is canceled after timeout?

hi @1996fanrui :

whenComplete is the completion of the call, run after future successful and exceptionally(Include timeout)

**Comment by wolfboys** [comment]
> Hi @wolfboys , just to managed the expectation, as I said before: `Sorry, I'm not familiar with flink on k8s related code, so I just feedback some technical comments.`
> 
> My comments are just code specification, I didn't review the core logical of this PR. So I cannot approve this PR after my comments are resolved. It still needs some k8s reviewers.
> 
> cc @MonsterChenzhuo

Thanks for your professional opinions. very good 👏👏👏

File: streampark-common/src/main/scala/org/apache/streampark/common/util/CompletableFutureUtils.scala
Comment (1996fanrui) [reply streampark-common/src/main/scala/org/apache/streampark/common/util/CompletableFutureUtils.scala]: I understand the code, I just mean we should add a unit test.

Comment (wolfboys) [reply streampark-common/src/main/scala/org/apache/streampark/common/util/CompletableFutureUtils.scala]: > I understand the code, I just mean we should add a unit test.

okay. I provide this testcase

File: streampark-common/src/main/scala/org/apache/streampark/common/util/CompletableFutureUtils.scala
Comment (1996fanrui) [reply streampark-common/src/main/scala/org/apache/streampark/common/util/CompletableFutureUtils.scala]: I know it can be executed multiple times, I mean we should not execute some useless code, and we just cancel the future when it needs to be canceled. 

I see you updated the `!future.isCanceled`, it doesn't make sense. When future is finished before timeout, we don't need cancel the future, however `future.isCanceled == false`.

That's why I said, we should check `!future.isDone` instead of `!future.isCanceled`

File: streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java
Comment (1996fanrui) [reply streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java]: You can take a look how to use the  `assertThatThrownBy`,  `assertThatThrownBy` will catch the exception and it can check exception type and exception message. Apache flink uses it as well.

This is flink community doc about junit5:

https://docs.google.com/document/d/1514Wa_aNB9bJUen4xm5uiuXOooOJTtXqS_Jqk9KJitU/edit#heading=h.we2n5aifcjfq

image

File: streampark-common/src/main/scala/org/apache/streampark/common/util/CompletableFutureUtils.scala
Comment (wolfboys) [reply streampark-common/src/main/scala/org/apache/streampark/common/util/CompletableFutureUtils.scala]: > I know it can be executed multiple times, I mean we should not execute some useless code, and we just cancel the future when it needs to be canceled.
> 
> I see you updated the `!future.isCanceled`, it doesn't make sense. When future is finished before timeout, we don't need cancel the future, however `future.isCanceled == false`.
> 
> That's why I said, we should check `!future.isDone` instead of `!future.isCanceled`

done

File: streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java
Comment (1996fanrui) [inline streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java]: The `org.apache` is repeated

File: streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java
Comment (1996fanrui) [inline streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java]:
```suggestion
    assertThatThrownBy(future::get)
        .cause()
        .as(expectedExceptionMessage)
        .isInstanceOf(RuntimeException.class);
```

**Review by 1996fanrui — CHANGES_REQUESTED** [review]

File: streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java
Comment (wolfboys) [reply streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java]: > 

Cannot resolve method 'cause' in 'AbstractThrowableAssert'

Comment (wolfboys) [reply streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java]: > .isInstanceOf(RuntimeException.class);

 assertThatThrownBy(future::get)
        .getCause()
        .as(expectedExceptionMessage)
        .isInstanceOf(RuntimeException.class);

**Comment by MonsterChenzhuo** [comment]
LGTM

**Review by MonsterChenzhuo — APPROVED** [review]

**Review by MonsterChenzhuo — APPROVED** [review]

File: streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java
Comment (wolfboys) [reply streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java]: > The `org.apache` is repeated

resolved

File: streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java
Comment (wolfboys) [reply streampark-console/streampark-console-service/src/test/java/org/apache/org/apache/streampark/common/util/CompletableFutureUtilsTest.java]: done

**Review by 1996fanrui — APPROVED** [review]
Thanks for your hard work, LGTM

**Pull request merged**
